### ISSUE TITLE
Virtual host support like in HTTP

### ISSUE THREAD

**Issue by mdevaev** [rfbproto/rfbproto#50]
In HTTP, it is possible to serve multiple virtual hosts on the same IP and port. To do this, the HTTP request headers specify the `Host`, which indicates which server should respond, like:
```
GET / HTTP/1.1
Host: example.com
...

```
We are faced with the need to do something similar for VNC, but unfortunately there is no way to differentiate the target server after the connection, in other words, in VNC, one server must always correspond to a dedicated port number. If it were not for this limitation, we could make a single entry point for hundreds of thousands of servers on the same IP, or even do load balancing and reverse proxying, like nginx.

I propose to somehow extend the protocol, or, if necessary, register a new version of it, which would allow specifying the fqdn of the virtual host. For example, we specify the hostname immediately after the version (connecting to `kvm1.example.com --> example.com` on the same IP).
```
=== Server ===
RFB 003.108\n

=== Client ===
RFB 003.108\n
kvm1.example.com\n
```

The bump of the version will not break the compatibility of the client with other servers, but for the server of the new version, clients that support this will be needed.

Discuss?

**Comment by RokerHRO** [comment]
Just use a decent IP version and you got thousands, if not millions of IP addresses for free. I won't add any protocol extension nor RFB protocol version just to support the limits of an outdated IP version.

**Comment by CendioOssman** [comment]
The version field is not a good extension point, as it is not very flexible. You can't sensibly combine different extensions there.

Have you considered a security type instead?

Or use an existing one? E.g. multiplex based on username?

Or TLS SNI?

**Comment by berrange** [comment]
The way most cloud / virtualization software works is to put an HTTP websockets proxy in front of the real VNC servers. So you can literally use existing HTTP virtual hosting functionality, and/or the URL paths to distinguish which backend VNC server the client is requesting to talk to, without requiring any RFB protocol extensions. There are of course various pros/cons to having a proxy like this, but it would be my default recommendation for anyone wanting to front multiple servers with a single ingress service.

**Comment by CendioOssman** [comment]
That would assume WebSocket support in the VNC client, though. I'm not sure what the current state of that is. TigerVNC doesn't have it, at least.

**Comment by mdevaev** [comment]
> Just use a decent IP version and you got thousands, if not millions of IP addresses for free. I won't add any protocol extension nor RFB protocol version just to support the limits of an outdated IP version.

It doesn't work in the real world. Look at the IPv6 availability graph.

**Comment by mdevaev** [comment]
> The version field is not a good extension point, as it is not very flexible. You can't sensibly combine different extensions there.

I don't quite understand why. The handshake implies that anything can go after the first line of the version, which is why I suggest placing it there.

> Have you considered a security type instead?

Do you mean to make a security type that will include other types, like vencrypt? Probably, this may well be the solution.

> Or use an existing one? E.g. multiplex based on username? Or TLS SNI?

Definitely not, in the case of encrypt and tls, the user is transmitted encrypted, and my goal is that each server can have its own valid certificate.


> WebSocket

I thought about it, but it looks like a special case for noVNC, and it's too difficult for the rest of the clients to support it too.

So far, the idea with the magic security type looks good.

**Comment by CendioOssman** [comment]
> > The version field is not a good extension point, as it is not very flexible. You can't sensibly combine different extensions there.
> 
> I don't quite understand why. The handshake implies that anything can go after the first line of the version, which is why I suggest placing it there.
> 

The problem is if I want to write a server that supports feature A, B and C. All of them indicate support by having a different version number. How do I advertise all three things?

**Comment by mdevaev** [comment]
Good point. Then we can make a sectype, which should be the only one in the list from the server, if the server only works in virtual hosting mode. If the client processes it and selects a host, the server can send the usual security types message and continue the handshake as usual, but already for the selected host.

**Comment by Neustradamus** [comment]
It is linked to:
- https://github.com/rfbproto/rfbproto/pull/51
